### PR TITLE
networkmanager: 1.30.4 -> 1.32.4

### PR DESCRIPTION
**Pull request by amaxine**
###### Motivation for this change

https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.32.4/NEWS
```markdown
===============================================
NetworkManager-1.32.4
Overview of changes since NetworkManager-1.32.2
===============================================

* core: remove stale entries from "seen-bssids" and "timestamp"
  files in "/var/lib/NetworkManager".
* bond: support the peer_notif_delay option.
* core: add ipv[46].required-timeout option to wait for IP
  configuration while activating.
* core: send ARP announcements when there is carrier.
* core: start DHCPv6 when a prefix delegation is needed for shared
  mode.
* firewall: fix nftables backend to create "ip" table for
  IPv4 only.
* initrd: set required-timeout of 20 seconds for default IPv4 configuration
  to opportunistically wait for IPv4.
* ifcfg: log warning about invalid keys in ifcfg files.
* ifcfg: reject non-UTF-8 from ifcfg files.
* nmcli: show DNS SEARCH field in device information.
* cloud-setup: add support for Aliyun cloud.


=============================================
NetworkManager-1.32.2
Overview of changes since NetworkManager-1.32
=============================================

* hostname: prefer IPv4 addresses for reverse DNS lookup.
* dhcp: ignore unauthenticated FORCERENEW messages with
  internal, systemd-based DHCPv4 plugin (CVE-2020-13529).
  This plugin is not used, unless the undocumented dhcp=systemd
  option was set.
* cloud-setup: preserve IP addresses, routes and rules from
  currently active connection profile.
* Various bugfixes and performance improvements.

=============================================
NetworkManager-1.32
Overview of changes since NetworkManager-1.30
=============================================

* Add an 'accept-all-mac-addresses' property to the ethernet setting
  to accept frames with any MAC address (also known as promiscuous
  mode).
* nmcli: fix setting property aliases to empty value to reset the
  default value.
* BEHAVIOR CHANGE: Enforce valid lower case UUID format for the "connection.uuid"
  by normalizing the value. This changes the UUID of existing profiles with an
  invalid or non-normalized value. Also normalize "connection.secondaries" to only
  contain valid and distinct UUIDs. Profiles that set "connection.master" or the
  "parent" properties to a non-normalized UUID need adjustment. Using such
  non-normalized UUIDs should be rare and affect few users.
* API CHANGE: D-Bus: remove long deprecated "PropertiesChanged" signal from D-Bus
  API. They are replaced by standard "PropertiesChanged" signal on the
  "org.freedesktop.DBus.Properties" interface. It is not expected that anybody
  is still using this API.
* Now NetworkManager uses systemd-resolved API to lookup the system
  hostname via reverse DNS. If systemd-resolved is not available, a
  'nm-daemon-helper' binary is spawned to perform the lookup using
  the 'dns' NSS module.
* dhcp: honor "ID_NET_DHCP_BROADCAST" udev attribute to set the broadcast flag.
  This allows to configure devices in udev for which DHCPOFFER messages are to be
  broadcast.
* firewall: add nftables firewall backend for configuring IPv4 NAT with shared
  mode. Now two backends are supported, "iptables" and "nftables". The default
  gets detected based on whether /usr/sbin/nft or /usr/sbin/iptables is installed,
  with nftables preferred.
  Please note that SELinux might prevent NetworkManager from communicating with nft.
* ethtool: add support for pause settings.
* Support "prohibit"/"blackhole"/"unreachable" type routing rules.
* Now NetworkManager preserves by default the existing traffic control
  configuration (qdiscs and filters) when activating a connection.
* wifi/iwd: mirror NetworkManager connection profiles to iwd config files.
* API CHANGE: vala: fix wrongly generated vapi for WireGuard API.
* wifi: fix parsing of Microsoft Network Cost information element for
  detecting metered networks.
* dhcp: support option 249 (Microsoft Classless Static Route) with nettools
  internal DHCP plugin.
* initrd: support the "rd.net.dhcp.retry" kernel argument for controlling retry
  and timeout to get a DHCP lease.
* API CHANGE: hide definitions of structs for NMSetting and NMSimpleConnection from
  public headers.
* Major restructuring of the source code.
* Many bugfixes and minor improvements.
```

~~Switched from iptables to nftables as it's upstream's preference. The paths patch needed rebasing again due to major restructuring.~~ When `nftables` is enabled, the NetworkManager module will now set the backend appropriately, additionally exposing the ability to disable this completely.

~~The `nss` dependency is from building `libnm-crypto-nss`, which after the restructuring gets built regardless of us using `gnutls`.~~ This is now resolved, with a patch already merged upstream.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by maxeaubrey** [comment]
Added a cleanup commit improving readability a bit (the only changes done not by `nixpkgs-fmt` is reodering of `mesonFlags`). Found parts of this to be kinda hard to read earlier.

**Review by endocrimes — APPROVED** [review]
Built by cherry-picking onto ToT and did basic validation of nm*, seems to be good.

Diff cleanup lgtm.

**Comment by maxeaubrey** [comment]
Bumped to 1.32.2

```
=============================================
NetworkManager-1.32.2
Overview of changes since NetworkManager-1.32
=============================================

* hostname: prefer IPv4 addresses for reverse DNS lookup.
* dhcp: ignore unauthenticated FORCERENEW messages with
  internal, systemd-based DHCPv4 plugin (CVE-2020-13529).
  This plugin is not used, unless the undocumented dhcp=systemd
  option was set.
* cloud-setup: preserve IP addresses, routes and rules from
  currently active connection profile.
* Various bugfixes and performance improvements.
```

I was running 1.32.0 for about ~10 days (I rolled back today to make keeping up updates easier) without issue, though obviously I didn't really extensively test all features.

**Comment by ajs124** [comment]
> The nss dependency is from building libnm-crypto-nss, which after the restructuring gets built regardless of us using gnutls.

If we're already depending on nss, can we maybe drop the gnutls dependency then?

**Comment by maxeaubrey** [comment]
> If we're already depending on nss, can we maybe drop the gnutls dependency then?

I'm not sure why we switched to gnutls in the first place. Iirc when I tried looking into it I didn't manage to find a specific reason, and other distributions vary in choice between nss/gnutls, without a strong upstream recommendation.

I have no strong opinions on the topic, but I didn't want to switch either way without knowing what the reasoning behind the original choice was :)

**Comment by ajs124** [comment]
I somehow doubt we're going to figure out the reason, seeing as this was committed with gnutls a decade ago: https://github.com/NixOS/nixpkgs/commit/9b50ebd825fe85c6cc20217b0740d8b7ed780d6a#diff-8078b679b3af734e5c1ded719c54a2ae80e04b99524c1ed90b2462703e0bf0d5R14

I don't really care either way, but depending on both seems kind of silly.

**Comment by maxeaubrey** [comment]
I don't disagree there, it would be best to fix this upstream.

However I don't feel comfortable pushing a change from gnutls to nss myself. Debian/Ubuntu and Fedora all build with gnutls instead, and I don't have enough experience or knowledge here to just casually push this change through - though I'd love to actually know what the potential side effects of this change are, if any.

**Comment by maxeaubrey** [comment]
Added a patch that no longer requires us to depend on both at build time. Not sure if it's the best way to do it, I'm not super experienced with meson, but it does the job.

I'll try opening an upstream MR with that patch too that can hopefully be in for the next release, provided what I'm doing there actually makes sense :sweat_smile:

**Comment by maxeaubrey** [comment]
I realised that defaulting to any one firewall backend here (even if in principle unlikely to cause breakage, afaiu) was a bad idea, and so now instead this change is set by the module, with iptables/nftables being set based on what module is enabled.

**Comment by maxeaubrey** [comment]
https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.32.4/NEWS
```
===============================================
NetworkManager-1.32.4
Overview of changes since NetworkManager-1.32.2
===============================================

* core: remove stale entries from "seen-bssids" and "timestamp"
  files in "/var/lib/NetworkManager".
* bond: support the peer_notif_delay option.
* core: add ipv[46].required-timeout option to wait for IP
  configuration while activating.
* core: send ARP announcements when there is carrier.
* core: start DHCPv6 when a prefix delegation is needed for shared
  mode.
* firewall: fix nftables backend to create "ip" table for
  IPv4 only.
* initrd: set required-timeout of 20 seconds for default IPv4 configuration
  to opportunistically wait for IPv4.
* ifcfg: log warning about invalid keys in ifcfg files.
* ifcfg: reject non-UTF-8 from ifcfg files.
* nmcli: show DNS SEARCH field in device information.
* cloud-setup: add support for Aliyun cloud.
```

And another release. This way I get to drop the patch introduced in the last bump as that's been upstreamed! :)

**Review by endocrimes — APPROVED** [review]
LGTM

**Review by SuperSandro2000 — APPROVED** [review]
I think this is fine but I am not fully sure on the module option and if this will cause many regressions or none at all.

**Comment by maxeaubrey** [comment]
> I think this is fine but I am not fully sure on the module option and if this will cause many regressions or none at all.

The only likely potential fallout to the change is for people who were using `nftables` already, were relying on the NAT related features in NetworkManager, and they were configuring the rules entirely by hand.

For most people this will be a no-op, I imagine there are very few people, if any, who will hit this.

**Comment by mweinelt** [comment]
Does that sound like it's worth a release notes entry? I'm not sure.

**Comment by maxeaubrey** [comment]
> Does that sound like it's worth a release notes entry? I'm not sure.

There isn't anything backwards incompatible about the module changes, disabling the `firewall-backend` setting (and therefore feature) currently required someone to override NetworkManager's managed configuration file, which while I don't doubt someone might have done, means they wouldn't notice this change at all anyway.

I am a bit concerned whether my use of `mkForce` in the `nftables` module is correct, but no one seems to have raised issue.

Is NetworkManager being updated to latest a "notable" change in this case? If so I can add a note mentioning the version change and referencing their [blog post](https://networkmanager.dev/blog/networkmanager-1-32/). We didn't do it for 21.05, but it would be nice to see effort done around improving release notes.

**Comment by mweinelt** [comment]
The primary change I was thinking about was that networkmanager now relies on nftables as its firewall backend. And yes, I stumbled upon your use of `mkForce` there. Is `mkDefault` not possible, that would leave some wiggle-room for endusers to do what ever they feel like.

File: nixos/modules/services/networking/networkmanager.nix [hunk 1]
```diff
@@ -244,6 +245,15 @@ in {
         '';
       };
 
+      firewallBackend = mkOption {
+        type = types.enum [ "iptables" "nftables" "none" ];
+        default = "iptables";
```
Comment (mweinelt) [inline nixos/modules/services/networking/networkmanager.nix:250]: And just now I'm wondering what the default prio for module options are. Like if they're `mkDefault` or `mkOverride` with a lower priority.

Comment (maxeaubrey) [reply nixos/modules/services/networking/networkmanager.nix:250]: If I understand `lib/modules.nix` correctly, `mkOptionDefault` is `1500` and `mkDefault` is `1000`, and the comments indicate that their current use is as intended. So I think that part is sorted now.

**Pull request merged**